src/core/debt_register.sol

```diff
@@ -107,79 +114,108 @@ contract DebtRegister is DSNote {
-        uint index_ = rdiv(latestRateIndex, index);
-        uint wad = rmul(debt_, index_) - debt_;
-        return (latestRateIndex, index_, wad);
+    // changes the interest rate of a rate category
+    function file(uint rate, uint speed_) public auth note {
+        require(speed_ != 0);
+        rates[rate].speed = speed_;
+        rates[rate].index = ONE;
+        rates[rate].rho = uint48(now);
+        drip(rate);
     }
 
-    // --- Rate Accumulation ---
-    function drip(uint rate) public {
-        if (now >= rates[rate].lastUpdated) {
-            (uint latestRateIndex, , uint wad) = compounding(rate);
-            rates[rate].index = latestRateIndex;
-            rates[rate].lastUpdated = uint48(now);
-            incTotalDebt(rate, wad);
-        }
+    // Converters:
+    // convert debt amount to chi
+    function toChi(uint index, uint wad) public view returns (uint) {
+        return rdiv(wad, index);
     }
 
-    function getCurrentDebt(uint loan, uint rate) public view returns (uint) {
-        uint index = rates[rate].index;
-        if (now >= rates[rate].lastUpdated) {
-            (index, ,) = compounding(rate);
-        }
-        return calcDebt(index, debtBalance[loan]);
+    // convert chi to debt amount
+    function fromChi(uint index, uint wad) public view returns (uint) {
+        return rmul(wad, index);
     }
 
-    function debtOf(uint loan, uint rate) public view returns(uint) {
-        return calcDebt(rates[rate].index, debtBalance[loan]);
-    }
+    // compound calculates the new index, the delta between old and new index and the delta of debt
+    function compound(uint rate) public view returns (uint index, uint index_delta, uint delta) {
+        uint48 rho = rates[rate].rho;
+        // require(now >= rho); TODO: why do we need to check for this edge case?
+        uint speed = rates[rate].speed;
+        uint index = rates[rate].index;
+        uint chi_ = rates[rate].chi;
+
+        // compounding in seconds
+        uint index_ = rmul(rpow(speed, now - rho, ONE), index);
+        uint debt = fromChi(chi_, index);
+        require(index != 0);
+        index_delta = rdiv(index_, index);
+        delta = rmul(chi_, index_) - debt;
 
-    function incTotalDebt(uint rate, uint wad) private {
-        rates[rate].debt = add(rates[rate].debt, wad);
-        totalDebt = add(totalDebt, wad);
+        return (index_, index_delta, delta);
     }
 
-    function decTotalDebt(uint rate, uint wad) private {
-        rates[rate].debt = sub(rates[rate].debt, wad);
-        totalDebt = sub(totalDebt, wad);
+    // drip updates the index of the rate category by compounding the interest and
+    // updates the total debt
+    function drip(uint rate) public {
+        if (now >= rates[rate].rho) {
+            (uint index, , uint delta) = compound(rate);
+            rates[rate].index = index;
+            rates[rate].rho = uint48(now);
+            total = add(total, delta);
```

Comment:
@manuelpolzhofer @ilinzweilin sorry for the confusion. This is actually why I changed to storing pie and not debt in the rate. If we do so, we save ourselves updating `Rate.pie` because it doesn't change if you call drip(). My changes I'm about to push will make this clearer.